### ISSUE TITLE
Adding reaction with new compartment does not create compartment

### ISSUE THREAD

**Issue by wbryant** [opencobra/cobrapy#257]
If I add a new reaction to a model, I can add metabolites from a new compartment (not in the current model), but the compartment is not created in the model.

I will have a go at fixing it myself when I have some time, but wondered if anyone has opinions on whether it should throw an error (i.e. assume compartments are incorrect), or whether a new compartment should (silently or not) be created in the model.

**Comment by aebrahim** [comment]
Thank you for reporting this

**Comment by pstjohn** [comment]
The `compartments` field in the model seems to be a holdover from SBML where the compartments are specified ahead of the metabolites.

I'd probably vote for removing it from `cobra.Model`, and have all of the `compartments` attribute for Reactions and Models be specified via properties:

```python
@property
def compartments(self):
    return {x.compartment for x in self._metabolites}
```

**Comment by mmundy42** [comment]
I recently ran across this issue too. It makes sense to remove `compartments` from the model and get the compartments from the metabolites. Although setting a compartment for a metabolite is optional and the default value is `None`. Will that cause any side effects?

**Comment by mmundy42** [comment]
Also, should there be a way to provide a name for a compartment ID?

**Comment by hredestig** [comment]
Fixed with #433
